Docker integ test with async API

normanj-bitquill opened the pull request:
### Description
Update the integration test docker stack to support the OpenSearch Async API and using Minio as an S3 storage engine. Also includes having everything configured on startup.

### Related Issues
#992 

### Check List
- [ ] Updated documentation (docs/ppl-lang/README.md)
- [ ] Implemented unit tests
- [ ] Implemented tests for combination with other commands
- [ ] New added source code should include a copyright header
- [ ] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

YANG-DB commented:
@normanj-bitquill I've also started working on some similar [PR with Iceberg based docker-compose](https://github.com/opensearch-project/opensearch-spark/pull/1001) - 
lets see if we can share knowledge ...

normanj-bitquill commented:
### Current Status

In my testing, I have added a second OpenSearch container. This resolves issues with the cluster and indices going into yellow state.

Changes in OpenSearch node:
* Bind the docker socket
* Replace the `aws-java-sdk-emrserverless-*.jar` file with on that uses a different EMRClient. The new EMRClient writes a file to a directory (`/tmp/docker`) with the arguments for running docker.
* A script is run on the container that reads files from `/tmp/docker` and runs docker with the arguments.
* Disable system indices and system indices permissions
* Connect with second OpenSearch container

What works:
* Initiate query with aysnc-api on OpenSearch container
* Starts a new Spark docker container to call `spark-submit` using information from the start job request that the EMRClient received
* New container connects to spark cluster to execute the job
* Creates the request and result indices
* Saves the result to the result index

What is missing:
* Detecting that the job has finished (this is on the spark-submit container)
* The query is for a table on an S3 data source, query fails with table not found

normanj-bitquill commented:
@YANG-DB I have partially working async API in the latest commit. These are my testing steps:

1. Start the cluster using `docker compose up`
2. Run `spark-shell` on the master spark container.
   1. Create an external table with a location under `s3a://integ-test/`
   2. Put some data in the table
3. Submit a query:
   ```
   curl -u '...' -X POST -H 'Content-Type: application/json' -d '{"datasource": "mys3", "lang": 
   "sql", "query": "SELECT * FROM mys3.default.foo"}
   ```
4. Retrieve the result (will need to wait until it is ready, maybe about 1 minute)
   ```
   curl -u '...' -X POST -H 'Content-Type: application/json' -d '{}' 'http://localhost:9200/query_execution_result_mys3/_search?pretty'
   ```

The OpenSearch container will need to bind the docker socket `/var/run/docker.sock`. By default this isn't available on Mac, but can be enabled.

The OpenSearch container will start another container to process the async query. This is the place where we could slip in the EMR spark container (if there is any value from it).

normanj-bitquill commented:
I have not tested retrieving results using the Async API. This is likely broken, since cannot check on the EMR job status. I also haven't tested a streaming query (also likely broken).

YANG-DB commented on docker/integ-test/docker-compose.yml:
```diff
@@ -8,29 +23,35 @@ services:
       - "${UI_PORT:-4040}:4040"
       - "${SPARK_CONNECT_PORT}:15002"
     entrypoint: /opt/bitnami/scripts/spark/master-entrypoint.sh
+    user: root
```
@normanj-bitquill why is this mandatory ?

normanj-bitquill replied:
Updated, the Spark master and worker nodes now use an image that is built from a `Dockerfile`. No need to start the containers with the `root` user.

YANG-DB commented on docker/integ-test/docker-compose.yml:
```diff
@@ -8,29 +23,35 @@ services:
-        source: $FLINT_JAR
+        source: ../../$FLINT_JAR
         target: /opt/bitnami/spark/jars/flint-spark-integration.jar
+      - type: bind
```
can you plz explain what is the issue here we need to solve ?

YANG-DB commented on docker/integ-test/opensearch/docker-command-runner.sh:
can you plz explain what is the issue here we need to solve ?

YANG-DB replied:
missing license header

normanj-bitquill replied:
I have explained it here.
https://github.com/opensearch-project/opensearch-spark/pull/1003#discussion_r1911887918

YANG-DB commented on docker/integ-test/opensearch/emr-src/com/amazonaws/services/emrserverless/AWSEMRServerlessClientBuilder.java:
```diff
@@ -0,0 +1,29 @@
+package com.amazonaws.services.emrserverless;
```
missing license header

YANG-DB commented on docker/integ-test/metastore/Dockerfile:
missing license header

YANG-DB commented on docker/integ-test/opensearch/Dockerfile:
missing license header

YANG-DB commented on docker/integ-test/opensearch/emr-src/org/opensearch/spark/emrserverless/DockerEMRServerlessClient.java:
missing license header

YANG-DB commented on docker/integ-test/opensearch/opensearch-docker-it-entrypoint.sh:
```diff
@@ -0,0 +1,11 @@
+
+su opensearch ./opensearch-docker-entrypoint.sh "$@"
+
+kill -TERM `cat /var/run/docker-command-runner.pid`
```
why is this called ?
can you plz explain the entire flow for the docker compose ?

YANG-DB replied:
ok nice !!

YANG-DB commented on docker/integ-test/spark/spark-master-entrypoint.sh:
```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
```
this is nice - can you add some documentation of what is done here ?

normanj-bitquill commented:
@YANG-DB Using the async API to get the results works once the results are available in the result index.

Some interesting behaviour though:
* When you submit a query, the container starts up and runs the query. It then has a loop (looking for new queries) and continues until it finally times out after 3 minutes.
* If you submit a query and then submit a second query (in a different session), the second query isn't actually run until the container for the first query has exited (after the 3 minute timeout)
* When the container starts up, it downloads the Flint and PPL extensions. This adds latency and I am trying to have it use the locally built Flint and PPL jar files.

I am likely missing some code for the new EMRClient for job management. This may explain some of the behaviour above. If the Spark app is going to run for 3 minutes, we should skip calling docker when the same session is used and the container is still running.

normanj-bitquill commented:
> @YANG-DB Using the async API to get the results works once the results are available in the result index.
> 
> Some interesting behaviour though:
> 
> * When you submit a query, the container starts up and runs the query. It then has a loop (looking for new queries) and continues until it finally times out after 3 minutes.
> * If you submit a query and then submit a second query (in a different session), the second query isn't actually run until the container for the first query has exited (after the 3 minute timeout)
> * When the container starts up, it downloads the Flint and PPL extensions. This adds latency and I am trying to have it use the locally built Flint and PPL jar files.
> 
> I am likely missing some code for the new EMRClient for job management. This may explain some of the behaviour above. If the Spark app is going to run for 3 minutes, we should skip calling docker when the same session is used and the container is still running.

I have fixed up this behaviour. The Spark cluster as is, only supports running one application at a time. I have changed the spark submit container to run the query locally. This allows multiple sessions to run at the same time, but each session could have its own container running.

I have also disabled downloading the dependencies on startup, to speed up query execution.

normanj-bitquill commented:
@YANG-DB I have added some documentation in this PR. Can you take a look and help with getting a second reviewer?

dai-chen reviewed (commented):
Is this IT with SQL async query API? Just wondering if this should be in SQL repo?

normanj-bitquill commented:
> Is this IT for Flint Spark with SQL async query API? Just wondering if this should be in SQL repo?

@dai-chen This is for three use cases:
* Submitting queries directly to Spark in order to test the PPL extension for Spark.
* Submitting queries directly to Spark that use the OpenSearch datasource. Useful for testing the Flint extension for Spark.
* Using the Async API to submit queries to the OpenSearch server. Useful for testing the EMR workflow and querying S3/Glue datasources.

Something very similar would make sense for the SQL repo. The SQL repo will not need the spark master and spark containers.

YANG-DB commented:
> > Is this IT for Flint Spark with SQL async query API? Just wondering if this should be in SQL repo?
> 
> @dai-chen This is for three use cases:
> 
> * Submitting queries directly to Spark in order to test the PPL extension for Spark.
> * Submitting queries directly to Spark that use the OpenSearch datasource. Useful for testing the Flint extension for Spark.
> * Using the Async API to submit queries to the OpenSearch server. Useful for testing the EMR workflow and querying S3/Glue datasources.
> 
> Something very similar would make sense for the SQL repo. The SQL repo will not need the spark master and spark containers.

Actually this make sense as a wide test harness for such use cases where we want to test connectivity between different opensearch datasources ...

penghuo commented on docker/integ-test/.env:
```diff
@@ -5,9 +5,13 @@ MASTER_UI_PORT=8080
 OPENSEARCH_PORT=9200
+OPENSEARCH_PA_PORT=9600
 OPENSEARCH_DASHBOARDS_PORT=5601
+S3_ACCESS_KEY=Vt7jnvi5BICr1rkfsheT
```
is it for minio? it could be anything string?

normanj-bitquill replied:
Yes, this is for minio container and is only valid for the minio container.

LantaoJin replied:
@penghuo the current end2end test always failed with
```
2025-05-12T01:57:02.7209031Z com.amazonaws.services.s3.model.AmazonS3Exception: The Access Key Id you provided does not exist in our records. (Service: Amazon S3; Status Code: 403; Error Code: InvalidAccessKeyId; Request ID: ...
```
What is the correct key is?

penghuo replied:
I do not know. it is not actually s3 key.

penghuo commented on docs/docker/integ-test/README.md:
```diff
@@ -0,0 +1,166 @@
+
+This container also has a docker volume used to persist the S3 data.
+
+### Configuration-Updated
```
typo. Configuration-updater?

penghuo commented on docs/docker/integ-test/README.md:
````diff
@@ -0,0 +1,166 @@
+docker compose down -d
+```
+
+## Creating Tables in S3
````
could add example of how to submit query using async API?

normanj-bitquill replied:
Added an example of submitting a query and getting the results.

penghuo commented on docs/docker/integ-test/README.md:
```diff
@@ -0,0 +1,166 @@
+1. Submitting queries directly to Spark in order to test the PPL extension for Spark.
+2. Submitting queries directly to Spark that use the OpenSearch datasource. Useful for testing the Flint extension
+   for Spark.
+3. Using the Async API to submit queries to the OpenSearch server. Useful for testing the EMR workflow and querying
```
what is the difference of 2 and 3?

normanj-bitquill replied:
For 2, a query is sent directly to Spark and Spark needs to query the index on the OpenSearch server. An example of where this might be useful is testing PPL on Spark, querying OpenSearch indices.

For 3, a query is sent directly to the OpenSearch server for the S3/Glue datasource. A new Spark container is started to process the request. The Spark container needs to query the table using Hive and minio. The Spark container is also making use of the `spark-sql-application` in this repository to process the query.

YANG-DB commented:
thanks @penghuo !
